# STAT 207 Group Lab Assignment 2 - [10 total points]

## Finding Missing Data and Answering Questions

<hr>

## <u>Lab Grading</u>:

Should we grade your submission?  If not, write the netID of the submission to be graded.  (Note: We will only grade one assignment per group, and we'll pick the first one that says we should grade that submission.  We will assign the same grade to all team members.)

*For example*, you might respond: **grade this submission** or **my submission is under netID jdeeke**

my submission is under netID rjguhl2

If you said **my submission is under netID** above, we will not read any more of your lab submission.

If you said **grade this submission** above, who worked with you on this submission?  Write both their **names** and **netIDs**.  

N/A

## <u>Purpose</u>:
You should work in groups of 2-3 on this report (not working in groups without permission will result in a point deduction). The purpose of this group lab assignment is to understand and wrangle data, so that it would be ready and prepared for future analysis.
<hr>

## <u>Assignment Instructions</u>:

### Group Roles

You are expected to work in groups of 2-3 on this report.  Since you are working in groups, you may find it helpful to have specified roles.  Below, I provide roles that can be used for groups of 2 and for groups of 3.  I encourage you to switch roles within this lab report, as possible.  I also encourage you to switch roles for each subsequent lab, as possible, based on your group membership.  

#### Groups of 2

* **Driver**: This student will type the report.  While typing the report, you may be the one who is selecting the functions to apply to the data.
* **Navigator**: This student will guide the process of answering the question.  Specific ways to help may include: outlining the general steps needed to solve a question (providing the overview), locating examples within the course notes, and reviewing each line of code as it is typed.

#### Groups of 3

* **Driver**: This student will type the report.  They may also be the one to select the functions to apply to the data.
* **Navigator**: This student will guide the process of answering the question.  They may select the general approach to answering the question and/or a few steps to be completed along the way. 
* **Communicator**: This student will review the report (as it is typed) to ensure that it is clear and concise.  This student may also locate relevant examples within the course notes that may help complete the assignment.

<hr>

### Imports

In [19]:
#Run this
import pandas as pd                    # imports pandas and calls the imported version 'pd'
import matplotlib.pyplot as plt        # imports the package and calls it 'plt'
import seaborn as sns                  # imports the seaborn package with the imported name 'sns'
sns.set()  

## Steam Data

Steam is the world's most popular PC Gaming hub. They have a massive catalog of games, with everything from AAA blockbusters to small indie titles.  

You are a team of data scientists working for Steam, and you are responsible for completing analysis for a report that will go to the executives of Valve (the parent company for Steam).  

Unfortunately, the data were collected with a less than optimal structure.  The dataset is comprised of transactions within the Steam platform for a random sample of 500 steam users along with their purchase and game play behaviors.  It has the following columns:
* user_id,
* game_name,
* activity:
    - purchase: indicating that the user has *purchased* the corresponding game
    - play: indicating that the user has *played* the corresponding game (for at least some amount of time.) 
* hours_played_if_play:
    - if the row corresponds to a 'play' activity, this number represents the number of hours the user has played the game
    - if the row corresponds to a 'purchase' activity, this number is always a 1 (and means nothing... it's a placeholder).

Note that for each user-game combo, there will either be one row (if the game has only been purchased but not played) or two rows (if the game has been both purchased and played).  For example, if I bought Portal, a row would be added to the data representing the purchase of that game by me.  When I open the game and start playing it, a second row would be added to the data representing that I have played the game along with recording my play time for that specific game.

### 1. [1 point] Reading in the Data

First, read the steam_sample.csv file into a dataframe.  Display the first five rows and the number of observations in the data.

Be sure to note the structure of the data (especially how the first two rows relate to each other) when observing the data.

In [21]:
df = pd.read_csv("steam_sample.csv")

In [22]:
df.head(10)

,user_id,game_name,activity,hours_played_if_play
0,308653033,Unturned,purchase,1
1,308653033,Unturned,play,0.6
2,308653033,theHunter,purchase,1
3,144004384,Dota 2,purchase,1
4,144004384,Dota 2,play,22
5,54103616,Counter-Strike Global Offensive,purchase,1
6,54103616,Counter-Strike Global Offensive,play,1028
7,54103616,Counter-Strike,purchase,1
8,54103616,Counter-Strike,play,1008
9,54103616,Left 4 Dead,purchase,1


### 2. [3 points] Preparing the Data

In order to prepare the data effectively, perform the following steps:

1. Identify any values (if any) that have been encoded in the csv to represent a **missing value**.  
2. Make sure that Python reads these missing values correctly.  
3. Report the number of rows that have missing values and the proportion of rows with missing values.
4. Drop any observations with missing values.

In [18]:
df.isnull().sum()

user_id                 0
game_name               0
activity                0
hours_played_if_play    0
dtype: int64

In [73]:
df[df['hours_played_if_play'] == 'unknown']

,user_id,game_name,activity,hours_played_if_play
322,135177440,Counter-Strike Global Offensive,purchase,unknown
323,135177440,Counter-Strike Global Offensive,play,unknown
358,110489310,Team Fortress 2,purchase,unknown
359,110489310,Team Fortress 2,play,unknown
380,106147454,Garry's Mod,purchase,unknown
381,106147454,Garry's Mod,play,unknown


In [72]:
df_cleaned = df[df['hours_played_if_play'] != 'unknown']
df_cleaned

,user_id,game_name,activity,hours_played_if_play
0,308653033,Unturned,purchase,1
1,308653033,Unturned,play,0.6
2,308653033,theHunter,purchase,1
3,144004384,Dota 2,purchase,1
4,144004384,Dota 2,play,22
...,...,...,...,...
7801,99096740,SimCity 4 Deluxe,play,0.2
7802,99096740,BioShock Infinite Burial at Sea - Episode 2,purchase,1
7803,99096740,The Elder Scrolls V Skyrim - Dawnguard,purchase,1
7804,99096740,The Elder Scrolls V Skyrim - Dragonborn,purchase,1


### 3. [1 point] Interpreting Missing Data

We dropped observations above that had any missing values.  Are you concerned that the decision to drop observations as the way to handle missing data may not have been best?  Briefly explain.

It's dependant on the number of missing values we dropped--if it's relatively huge, then it may affect our analysis negatively because we have smaller sample size, but if it's substantially small, then it shouldn't have too great of an affect.

### 4. [1 point] Separate the Data

In this analysis, we would like to answer questions based on purchases and based on play time.  Create two new dataframes:

- one that is comprised of the purchase rows for the games that were purchased
- one that is comprised of the play rows for the games that were played

You can call these anything that you want, but one option is to call them df_purchase and df_play.

In [76]:
df_purchase = df_cleaned[df_cleaned['activity'] == 'purchase']
df_purchase

,user_id,game_name,activity,hours_played_if_play
0,308653033,Unturned,purchase,1
2,308653033,theHunter,purchase,1
3,144004384,Dota 2,purchase,1
5,54103616,Counter-Strike Global Offensive,purchase,1
7,54103616,Counter-Strike,purchase,1
...,...,...,...,...
7800,99096740,SimCity 4 Deluxe,purchase,1
7802,99096740,BioShock Infinite Burial at Sea - Episode 2,purchase,1
7803,99096740,The Elder Scrolls V Skyrim - Dawnguard,purchase,1
7804,99096740,The Elder Scrolls V Skyrim - Dragonborn,purchase,1


In [77]:
df_play = df_cleaned[df_cleaned['activity'] == 'play']
df_play

,user_id,game_name,activity,hours_played_if_play
1,308653033,Unturned,play,0.6
4,144004384,Dota 2,play,22
6,54103616,Counter-Strike Global Offensive,play,1028
8,54103616,Counter-Strike,play,1008
10,54103616,Left 4 Dead,play,148
...,...,...,...,...
7793,99096740,Crysis,play,5.3
7795,99096740,Assassin's Creed II,play,2.7
7797,99096740,Hitman Blood Money,play,1.3
7799,99096740,The Binding of Isaac Rebirth,play,0.7


### 5. [2 points] Purchasing to Playing?

Overall, of all of the purchases represented in this data, what proportion have been played?  Of all the purchases in the data, what proportion remain "unopened", that is, unplayed?

What factors do you anticipate might affect whether a given game is played or not after being purchased?

In [78]:
(len(df_purchase) / len(df)) * 100

63.13092492954138

In [79]:
(len(df_play) / len(df)) * 100

36.792211119651554

63.1% have been played.
36.8% remain "unopened" or unplayed.

### 6. [2 points] Game Play Time Question

**a)** Review the amount of time played for each row (user-game combination).  You will want to summarize the time played variable.  You may also want to explore some of the extreme tails (very small or very large values) using visualizations and/or examining the specific values.

In [101]:
df_hours_smaller_than_20min = df_cleaned[df_cleaned['hours_played_if_play'] < '0.2']
df_hours_smaller_than_20min

,user_id,game_name,activity,hours_played_if_play
170,214388411,Realm of the Mad God,play,0.1
568,257422299,Metal Reaper Online,play,0.1
570,257422299,Golden Rush,play,0.1
668,35729292,Trove,play,0.1
724,185914106,Velvet Sundown,play,0.1
...,...,...,...,...
7157,75598925,The Showdown Effect,play,0.1
7175,165906966,Dota 2,play,0.1
7182,186123132,Unturned,play,0.1
7603,163968268,School of Dragons How to Train Your Dragon,play,0.1


In [110]:
df_cleaned.max()

user_id                 309265377
game_name               theHunter
activity                 purchase
hours_played_if_play          997
dtype: object

In [111]:
summary_stats = df_cleaned['hours_played_if_play'].describe()
summary_stats

count     7800
unique     472
top          1
freq      4976
Name: hours_played_if_play, dtype: object

**b)** Are there any of these values that you suspect may not be accurate?  Explain.  